compiler/rustc_mir_dataflow/src/impls/liveness.rs

```diff
@@ -179,3 +183,115 @@ impl DefUse {
         }
     }
 }
+
+/// Like `MaybeLiveLocals`, but does not mark locals as live if they are used in a dead assignment.
+///
+/// This is basically written for dead store elimination and nothing else.
+///
+/// All of the caveats of `MaybeLiveLocals` apply.
+pub struct MaybeTransitiveLiveLocals<'a>(&'a BitSet<Local>);
+
+impl<'a> MaybeTransitiveLiveLocals<'a> {
+    /// The `always_alive` set is the set of locals to which all stores should unconditionally be
+    /// considered live.
+    ///
+    /// This should include at least all locals that are ever borrowed.
+    pub fn new(always_alive: &'a BitSet<Local>) -> Self {
+        MaybeTransitiveLiveLocals(always_alive)
+    }
+}
+
+impl<'a, 'tcx> AnalysisDomain<'tcx> for MaybeTransitiveLiveLocals<'a> {
+    type Domain = BitSet<Local>;
+    type Direction = Backward;
+
+    const NAME: &'static str = "transitive liveness";
+
+    fn bottom_value(&self, body: &mir::Body<'tcx>) -> Self::Domain {
+        // bottom = not live
+        BitSet::new_empty(body.local_decls.len())
+    }
+
+    fn initialize_start_block(&self, _: &mir::Body<'tcx>, _: &mut Self::Domain) {
+        // No variables are live until we observe a use
+    }
+}
+
+struct TransferWrapper<'a>(&'a mut BitSet<Local>);
+
+impl<'a> GenKill<Local> for TransferWrapper<'a> {
+    fn gen(&mut self, l: Local) {
+        self.0.insert(l);
+    }
+
+    fn kill(&mut self, l: Local) {
+        self.0.remove(l);
+    }
+}
+
+impl<'a, 'tcx> Analysis<'tcx> for MaybeTransitiveLiveLocals<'a> {
+    fn apply_statement_effect(
+        &self,
+        trans: &mut Self::Domain,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    ) {
+        // Compute the place that we are storing to, if any
+        let destination = match &statement.kind {
+            StatementKind::Assign(assign) => Some(assign.0),
+            StatementKind::SetDiscriminant { place, .. }
+            | StatementKind::Retag(_, place)
```

Comment:
Is a place in a `Retag` statement a store destination? (I don't know how it is used by miri).

Reply:
A `Retag` is a read-modify-write. It is also impure (because it has side-effects on the borrow stack), but I *thought* that this should be not a concern. I'm realizing now I was wrong though.

@RalfJung I would like the following theorem: If `_5` is dead immediately after `Retag(_5)`, then it is sound to remove the `Retag`. As far as I can see this should be true with raw pointer tagging (I have no proof though), but is [false without raw pointer tagging](https://github.com/rust-lang/unsafe-code-guidelines/issues/273). For now, I'll change the implementation here to consider `Retag` to be a full side-effect, but is it worth tracking this elsewhere? It seems sort of important

Reply:
> I would like the following theorem: If _5 is dead immediately after Retag(_5), then it is sound to remove the Retag. 

That is certainly the intention, but we have not proven such a theorem.